drivers/modem/hl7800.c

```diff
@@ -0,0 +1,4902 @@
+		}
+
+		LOG_DBG("OUT: [%s]", log_strdup(data));
+		mdm_receiver_send(&ictx.mdm_ctx, data, strlen(data));
+		mdm_receiver_send(&ictx.mdm_ctx, "\r", 1);
+
+		if (K_TIMEOUT_EQ(timeout, K_NO_WAIT)) {
+			goto done;
+		}
+
+		if (!sock) {
+			ret = k_sem_take(&ictx.response_sem, timeout);
+		} else {
+			ret = k_sem_take(&sock->sock_send_sem, timeout);
+		}
+
+		if (ret == 0) {
+			ret = ictx.last_error;
+		} else if (ret == -EAGAIN) {
+			ret = -ETIMEDOUT;
+		}
+
+		retries--;
+		if (retries < 0) {
+			retries = 0;
+		}
+	} while (ret != 0 && retries > 0);
+done:
+	ictx.search_no_id_resp = false;
+	return ret;
+}
+
+static int wakeup_hl7800(void)
+{
+#ifdef CONFIG_MODEM_HL7800_LOW_POWER_MODE
+	int ret;
+
+	allow_sleep(false);
+	if (!is_cmd_ready()) {
+		LOG_DBG("Waiting to wakeup");
+		ret = k_sem_take(&ictx.mdm_awake, MDM_WAKEUP_TIME);
+		if (ret) {
+			LOG_DBG("Err waiting for wakeup: %d", ret);
+		}
+	}
+#endif
+	return 0;
+}
+
+int32_t mdm_hl7800_send_at_cmd(const uint8_t *data)
+{
+	int ret;
+
+	if (!data) {
+		return -EINVAL;
+	}
+
+	hl7800_lock();
+	wakeup_hl7800();
+	ictx.last_socket_id = 0;
+	ret = send_at_cmd(NULL, data, MDM_CMD_SEND_TIMEOUT, 0, false);
+	allow_sleep(true);
+	hl7800_unlock();
+	return ret;
+}
+
+/* The access point name (and username and password) are stored in the modem's
+ * non-volatile memory.
+ */
+int32_t mdm_hl7800_update_apn(char *access_point_name)
+{
+	int ret = -EINVAL;
+
+	hl7800_lock();
+	wakeup_hl7800();
+	ictx.last_socket_id = 0;
+	ret = write_apn(access_point_name);
+	allow_sleep(true);
+	hl7800_unlock();
+
+	if (ret >= 0) {
+		/* After a reset the APN will be re-read from the modem
+		 * and an event will be generated.
+		 */
+		k_delayed_work_submit_to_queue(&hl7800_workq,
+					       &ictx.mdm_reset_work, K_NO_WAIT);
+	}
+	return ret;
+}
+
+bool mdm_hl7800_valid_rat(uint8_t value)
+{
+	if ((value == MDM_RAT_CAT_M1) || (value == MDM_RAT_CAT_NB1)) {
+		return true;
+	}
+	return false;
+}
+
+int32_t mdm_hl7800_update_rat(enum mdm_hl7800_radio_mode value)
+{
+	int ret = -EINVAL;
+
+	if (value == ictx.mdm_rat) {
+		/* The set command will fail (in the modem)
+		 * if the RAT isn't different.
+		 */
+		return 0;
+	} else if (!mdm_hl7800_valid_rat(value)) {
+		return ret;
+	}
+
+	hl7800_lock();
+	wakeup_hl7800();
+	ictx.last_socket_id = 0;
+
+	if (value == MDM_RAT_CAT_M1) {
+		SEND_AT_CMD_ONCE_EXPECT_OK("AT+KSRAT=0");
+	} else { /* MDM_RAT_CAT_NB1 */
+		SEND_AT_CMD_ONCE_EXPECT_OK("AT+KSRAT=1");
+	}
+
+error:
+	if (ret >= 0) {
+		/* Changing the RAT causes the modem to reset. */
+		ret = modem_boot_handler("RAT changed");
+	}
+
+	allow_sleep(true);
+	hl7800_unlock();
+
+	/* A reset and reconfigure ensures the modem configuration and
+	 * state are valid
+	 */
+	if (ret >= 0) {
+		k_delayed_work_submit_to_queue(&hl7800_workq,
+					       &ictx.mdm_reset_work, K_NO_WAIT);
+	}
+
+	return ret;
+}
+
+#ifdef CONFIG_NEWLIB_LIBC
+int32_t mdm_hl7800_get_local_time(struct tm *tm, int32_t *offset)
```

Comment:
How is this used?

Reply:
We have a public API so the user can get the time from the modem in their app if they wish.